File: Leetcode 4001-4100/4001/4001.py
# Leetcode 4001: Aggregate Two Time Series
# https://leetcode.com/problems/aggregate-two-time-series/
# Solved on 23rd of August, 2026
class Solution:
    def aggregateTimeSeries(self, series1: list[list[int]], series2: list[list[int]]) -> list[list[int]]:
        """
        Aggregates two time series by merging them at common timestamps and taking the 
        latest value for each time series at unique timestamps.
        @param series1 First time series as a 2D array of [timestamp, value].
        @param series2 Second time series as a 2D array of [timestamp, value].
        @return The aggregated time series as a 2D array of [timestamp, value].
        """
        iOne = len(series1) - 1
        iTwo = len(series2) - 1
        vOne = 0
        vTwo = 0
        aSeries = []

        while iOne >= 0 or iTwo >= 0:
            if iOne >= 0 and iTwo >= 0:
                timeOne = series1[iOne][0]
                timeTwo = series2[iTwo][0]

                if timeOne > timeTwo:
                    currentTime = timeOne
                    vOne = series1[iOne][1]
                    iOne -= 1
                elif timeTwo > timeOne:
                    currentTime = timeTwo
                    vTwo = series2[iTwo][1]
                    iTwo -= 1
                else:
                    currentTime = timeOne
                    vOne = series1[iOne][1]
                    vTwo = series2[iTwo][1]
                    iOne -= 1
                    iTwo -= 1
            elif iOne >= 0:
                currentTime = series1[iOne][0]
                vOne = series1[iOne][1]
                iOne -= 1
            else:
                currentTime = series2[iTwo][0]
                vTwo = series2[iTwo][1]
                iTwo -= 1

            aSeries.append([currentTime, vOne + vTwo])

        aSeries.reverse()
        return aSeries

File: Leetcode 4001-4100/4001/4001.java
// Leetcode 4001: Aggregate Two Time Series
// https://leetcode.com/problems/aggregate-two-time-series/ 
// Solved on 23rd of August, 2026
class Solution {
    /**
     * Aggregates two time series by merging them at common timestamps and taking the 
     * latest value for each time series at unique timestamps.
     * @param series1 First time series as a 2D array of [timestamp, value].
     * @param series2 Second time series as a 2D array of [timestamp, value].
     * @return The aggregated time series as a 2D array of [timestamp, value].
     */
    public List<List<Integer>> aggregateTimeSeries(int[][] series1, int[][] series2) {
        List<List<Integer>> result = new ArrayList<>();
        int index1 = series1.length - 1;
        int index2 = series2.length - 1;
        int latestVal1 = 0;
        int latestVal2 = 0;

        while (index1 >= 0 || index2 >= 0) {
            int currentTimestamp;
            if (index1 >= 0 && index2 >= 0) {
                if (series1[index1][0] > series2[index2][0]) {
                    currentTimestamp = series1[index1][0];
                    latestVal1 = series1[index1][1];
                    index1--;
                } else if (series1[index1][0] < series2[index2][0]) {
                    currentTimestamp = series2[index2][0];
                    latestVal2 = series2[index2][1];
                    index2--;
                } else {
                    currentTimestamp = series1[index1][0];
                    latestVal1 = series1[index1][1];
                    latestVal2 = series2[index2][1];
                    index1--;
                    index2--;
                }
            } else if (index1 >= 0) {
                currentTimestamp = series1[index1][0];
                latestVal1 = series1[index1][1];
                index1--;
            } else {
                currentTimestamp = series2[index2][0];
                latestVal2 = series2[index2][1];
                index2--;
            }

            result.add(Arrays.asList(currentTimestamp, latestVal1 + latestVal2));
        }

        Collections.reverse(result);
        return result;
    }
}